### DISCUSSION TITLE
What is the proper way to solve "Document is not defined" error?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#36838] [Help]
I am working on converting this site to the Next.js platform, and I am getting this error:

```
ReferenceError: document is not defined
  141 | 
  142 |   // Force
> 143 |   let forceResult = document.getElementById("forceResult");
      |                                  ^
  144 | 
  145 |   function force1(pressure, area) {
  146 |     return pressure * area;
```

(see attached picture)

![error](https://user-images.githubusercontent.com/93617433/167899822-02ff0456-b73d-44d0-8e1a-638f7a9720b8.PNG)

What would be the current proper way to fix it for Next.js? 

If you want to see the repo code is below... I need to fix like about 12 codes using the same format... 
https://github.com/Elindo586/nextprojecta/blob/main/pages/fluid-power-formulas.js

**Comment by Elindo586** [comment]
This was fixed by including 
`if (typeof window !== "undefined) {}`

like this:

```
 if (typeof window !== "undefined") {
    let forceResult = document.getElementById("forceResult");
  }

  function force1(pressure, area) {
    return pressure * area;
  }

  const equals = () => {
    const pressureValue = Number(isPressure);
    const areaValue = Number(isArea);
    const result = force1(pressureValue, areaValue);

    forceResult.innerHTML = result + " lbs";
};
```
  If anybody has a better way let me know!!  thanks!

**Reply by icyJoseph** [reply]
This is a good way, but not the best way, though, it helps you understand, that, during rendering on the server, there's no window/document, or HTML elements at all. The server outputs a string.

That being said, here specifically, you should use React's `refs`!

```jsx
            <output type="number" id="forceResult" ref={forceResultRef}></output>
```

Keep the id if you wish, but if you create `const forceResultRef = useRef();`, you can then do `forceResultRef.current`, and that'll give you same as `getElementById`, in a more idiomatic React way, that doesn't require all of these window guards.

```js
  const equals = () => {
    const pressureValue = Number(isPressure);
    const areaValue = Number(isArea);
    const result = force1(pressureValue, areaValue);

    forceResult.current.innerHTML = result + " lbs";
  };
```

Though! I would also recommend to avoid `innerHTML`, it looks to me like there's a bunch of anti patterns in these components, but a first would be to use `forceResult.current.textContent = result + "lbs"`

> What is the proper way to solve "Document is not defined" error?

In this case, use refs!

--- 

I don't mean this to turn into a lesson about Refs and HTMLElements, but having a framework that gave no access to the DOM nodes it creates, would be very difficult to work with, and as you see, creates complex code. That's where passing Refs to the JSX comes in! In React, using selectors to get references to the DOM elements we define, is very rare. If you are in control of the DOM elements, use Refs.
